File: src/utils/Time.cc
#include "Time.h"

#include <chrono>
#include <ctime>

namespace ccc {
    namespace {
        Time s_time;
    }

    void Time::init() {
        data.start_time = data.now_time = data.last_time = std::chrono::steady_clock::now().time_since_epoch().count();
        data.delta_time = data.total_time = 0;
    }

    void Time::tick() {
        data.last_time = data.now_time;
        data.now_time = std::chrono::steady_clock::now().time_since_epoch().count();
        data.delta_time_raw = data.now_time - data.last_time;
        data.total_time_raw = data.now_time - data.start_time;
        data.delta_time = std::chrono::duration<double>(time::delta_raw()).count();
        data.total_time = std::chrono::duration<double>(time::total_raw()).count();
    }

    double Time::delta() const {
        return data.delta_time;
    }

    double Time::total() const {
        return data.total_time;
    }

    TimeData *time::get_data_ptr() {
        return &s_time.data;
    }

    void time::init() {
        s_time.init();
    }

    void time::tick() {
        s_time.tick();
    }

    std::chrono::steady_clock::duration time::delta_raw() {
        return std::chrono::steady_clock::duration(s_time.data.delta_time_raw);
    }

    std::chrono::steady_clock::duration time::total_raw() {
        return std::chrono::steady_clock::duration(s_time.data.total_time_raw);
    }

    double time::delta() {
        return s_time.delta();
    }

    double time::total() {
        return s_time.total();
    }
} // ccc

File: Dotnet/Soario/Utils/Time.cs
using System.Runtime.CompilerServices;
using Soario.Native;

namespace Soario.Utils;

public static unsafe class Time
{
    internal static TimeData* p_time_data;

    /// <summary>
    /// 开始执行时的单调时间
    /// </summary>
    public static TimeSpan StartTime
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => new(p_time_data->start_time / 100);
    }
    
    /// <summary>
    /// 上帧的单调时间
    /// </summary>
    public static TimeSpan LastTime
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => new(p_time_data->last_time / 100);
    }
    
    /// <summary>
    /// 当前帧的单调时间
    /// </summary>
    public static TimeSpan NowTime
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => new(p_time_data->now_time / 100);
    }

    /// <summary>
    /// 时间增量
    /// </summary>
    public static double Delta
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => p_time_data->delta_time;
    }

    /// <summary>
    /// 距离开始执行过去了多久
    /// </summary>
    public static double Total
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => p_time_data->total_time;
    }

    /// <summary>
    /// 时间增量
    /// </summary>
    public static TimeSpan DeltaRaw
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => new(p_time_data->delta_time_raw / 100);
    }

    /// <summary>
    /// 距离开始执行过去了多久
    /// </summary>
    public static TimeSpan TotalRaw
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        get => new(p_time_data->total_time_raw / 100);
    }
}
